### PR TITLE
ITSASU-506: update the update route in edit mode and back link

### PR DESCRIPTION
**Pull request by scullythornton1**

### CODE REVIEW

File: app/uk/gov/hmrc/incometaxsubscriptionselfemployedfrontend/controllers/BusinessNameController.scala [hunk 1]
```diff
@@ -79,15 +79,11 @@ class BusinessNameController @Inject()(mcc: MessagesControllerComponents,
...
-                if (isEditMode) {
-                  Redirect(uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.BusinessListCYAController.show())
-                } else {
-                  Redirect(uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.BusinessTradeNameController.show(id))
-                }
+              (isEditMode, isSaveAndRetrieve) match {
+                case (true, true) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.SelfEmployedCYAController.show(id).url
+                case (false, true) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.BusinessStartDateController.show(id)
+                case (true, false) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.BusinessListCYAController.show().url
+                case (false, false) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.BusinessTradeNameController.show(id).url
```
Comment (sb-dev) [inline app/uk/gov/hmrc/incometaxsubscriptionselfemployedfrontend/controllers/BusinessNameController.scala:86]: There are few issues with this function:
```
              multipleSelfEmploymentsService.saveBusinessTrade(id, businessTradeNameData).map(_ => {
                val call = (isEditMode, isSaveAndRetrieve) match {
                  case (true, true) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.SelfEmployedCYAController.show(id)
                  case (false, true) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.AddressLookupRoutingController.initialiseAddressLookupJourney(id)
                  case (true, false) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.BusinessListCYAController.show()
                  case (false, false) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.AddressLookupRoutingController.initialiseAddressLookupJourney(id)
                }
                Redirect(call)
              })
```

Comment (scullythornton1) [reply app/uk/gov/hmrc/incometaxsubscriptionselfemployedfrontend/controllers/BusinessNameController.scala:86]: hi Samir, I think the logic I put is correct. when is not editMode but saveAndRetrieve is enabled, the page after the business name should be the business start date. 
and when it is not in editMode and disabled saveAndRetrieve, the next page from buinsessName should be business trade.

Comment (sb-dev) [reply app/uk/gov/hmrc/incometaxsubscriptionselfemployedfrontend/controllers/BusinessNameController.scala:86]: Sorry, the logic was wrong in my comment. I added this comment to show you that the pattern matching should return a Call in order to redirect he user.

File: app/uk/gov/hmrc/incometaxsubscriptionselfemployedfrontend/controllers/BusinessTradeNameController.scala [hunk 2]
```diff
@@ -75,13 +77,18 @@ class BusinessTradeNameController @Inject()(mcc: MessagesControllerComponents,
...
-                    uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.AddressLookupRoutingController.initialiseAddressLookupJourney(id))
+              multipleSelfEmploymentsService.saveBusinessTrade(id, businessTradeNameData).map(_ => {
+                val call = (isEditMode, isSaveAndRetrieve) match {
+                  case (true, true) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.SelfEmployedCYAController.show(id)
+                  case (false, true) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.AddressLookupRoutingController.initialiseAddressLookupJourney(id)
+                  case (true, false) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.BusinessListCYAController.show()
+                  case (false, false) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.AddressLookupRoutingController.initialiseAddressLookupJourney(id)
                 }
+
+                println("isEditMode: " + isEditMode)
```
Comment (sb-dev) [inline app/uk/gov/hmrc/incometaxsubscriptionselfemployedfrontend/controllers/BusinessTradeNameController.scala:88]: I think you forgot to remove this one :)

File: it/controllers/BusinessTradeNameControllerISpec.scala [hunk 3]
```diff
@@ -41,132 +48,182 @@ class BusinessTradeNameControllerISpec extends ComponentSpecBase {
...
+//        val res = submitBusinessTradeName(businessId, inEditMode = false, Some(testInvalidBusinessTradeNameModel))
+//
+//        Then("Should return a BAD_REQUEST and THE FORM With errors")
+//        res must have(
+//          httpStatus(BAD_REQUEST),
+//          pageTitle("Error: What is the trade of your business?" + titleSuffix)
+//        )
+//      }
+//
+//    }
```
Comment (sb-dev) [inline it/controllers/BusinessTradeNameControllerISpec.scala:185]: Please uncomment these tests

File: app/uk/gov/hmrc/incometaxsubscriptionselfemployedfrontend/controllers/BusinessTradeNameController.scala (hunk 2 above)
Comment (scullythornton1) [reply app/uk/gov/hmrc/incometaxsubscriptionselfemployedfrontend/controllers/BusinessTradeNameController.scala:88]: sorry, :) will do it

File: test/uk/gov/hmrc/incometaxsubscriptionselfemployedfrontend/controllers/BusinessTradeNameControllerSpec.scala [hunk 4]
```diff
@@ -186,73 +210,63 @@ class BusinessTradeNameControllerSpec extends ControllerBaseSpec
...
-        status(result) mustBe SEE_OTHER
-        redirectLocation(result) mustBe Some(routes.AddressLookupRoutingController.initialiseAddressLookupJourney(id).url)
-      }
-      }
+
+      status(result) mustBe SEE_OTHER
+      redirectLocation(result) mustBe Some(routes.BusinessListCYAController.show().url)
+    }
     }
   }
```
[the remark this answers is not in the document]
Comment (scullythornton1) [reply test/uk/gov/hmrc/incometaxsubscriptionselfemployedfrontend/controllers/BusinessTradeNameControllerSpec.scala:272]: I have got that test row 215-234

File: it/controllers/BusinessTradeNameControllerISpec.scala (hunk 3 above)
Comment (scullythornton1) [reply it/controllers/BusinessTradeNameControllerISpec.scala:185]: uncomomented

File: test/uk/gov/hmrc/incometaxsubscriptionselfemployedfrontend/controllers/BusinessTradeNameControllerSpec.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (sb-dev) [reply test/uk/gov/hmrc/incometaxsubscriptionselfemployedfrontend/controllers/BusinessTradeNameControllerSpec.scala:272]: Sorry, I meant a test to cover the `backUrl` function when isEditMode = true and SaveAndRetrieve = true.

**Review by sb-dev — APPROVED** [review]

**Pull request merged**